src/transaction_builder.cpp

```diff
@@ -15,6 +15,81 @@
 #include <librustzcash.h>
 #include <rust/ed25519.h>
 
+uint256 ProduceZip244SignatureHash(
+    const CTransaction& tx,
+    const orchard::UnauthorizedBundle& orchardBundle)
+{
+    uint256 dataToBeSigned;
+    PrecomputedTransactionData local(tx);
```

Comment:
There is logical overlap: the hashes precomputed on the C++ side are identical to ones precomputed on the Rust side. The only reason for the internal split is that we've so far only focused on using ZIP 244 on the Rust side, but at some point I'd like to move our impls of ZIP 143 and 243 to also use Rust canonically (I've opened #5556 for this). At that point, this would revert to being a single-nature type. Thus I don't see the benefit of refactoring this to `std::variant` now, only to undo that refactor later.